src/script/interpreter.cpp

```diff
@@ -591,7 +593,54 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
                     break;
                 }
 
-                case OP_NOP1: case OP_NOP4: case OP_NOP5:
+                case OP_TXHASH: case OP_CHECKTXHASHVERIFY:
+                {
+                    // if flags not enabled; treat as a NOP4
+                    if (!(flags & SCRIPT_VERIFY_TXHASH)) {
+                        if (flags & SCRIPT_VERIFY_DISCOURAGE_UPGRADABLE_NOPS)
+                            return set_error(serror, SCRIPT_ERR_DISCOURAGE_UPGRADABLE_NOPS);
+                        break;
+                    }
+
+                    if (stack.size() < 1) {
+                        return set_error(serror, SCRIPT_ERR_INVALID_STACK_OPERATION);
+                    }
+
+                    Span<const unsigned char> tx_field_selector { stack.back() };
+                    if (opcode == OP_CHECKTXHASHVERIFY) {
+                        if (stack.back().size() < 32) {
+                            return set_error(serror, SCRIPT_ERR_CHECKTXHASHVERIFY);
+                        }
+                        tx_field_selector = tx_field_selector.subspan(32);
+                    }
+
+                    if (sigversion == SigVersion::TAPSCRIPT) {
+                        assert(execdata.m_validation_weight_left_init);
+                        execdata.m_validation_weight_left -= VALIDATION_WEIGHT_PER_TXHASH;
+                        if (execdata.m_validation_weight_left < 0) {
+                            return set_error(serror, SCRIPT_ERR_TAPSCRIPT_VALIDATION_WEIGHT);
+                        }
+                    }
+
+                    uint256 txhash;
+                    if (!checker.CalculateTxHash(txhash, tx_field_selector, execdata)) {
+                        return set_error(serror, SCRIPT_ERR_INVALID_TXFIELDSELECTOR);
+                    }
+
+                    if (opcode == OP_CHECKTXHASHVERIFY) {
+                        if (!memcmp(txhash.begin(), stack.back().data(), 32)) {
+                            return set_error(serror, SCRIPT_ERR_CHECKTXHASHVERIFY);
+                        }
+                        // nothing more to do, leave item on stack
```

Comment:
Leaving it on the stack seems inconsistent with the semantics of the other 'VERIFY' opcodes